calyx/src/passes/register_unsharing.rs

```diff
@@ -0,0 +1,287 @@
+use crate::analysis::reaching_defns::{
+    GroupOrInvoke, ReachingDefinitionAnalysis, INVOKE_PREFIX,
+};
+use crate::ir::traversal::{Action, Named, VisResult, Visitor};
+use crate::ir::{self, Builder, Cell, Group, LibrarySignatures, RRC};
+use std::collections::HashMap;
+use std::rc::Rc;
+
+#[derive(Default)]
+pub struct RegisterUnsharing {
+    // This is an option because it cannot be initialized until the component is
+    // seen
+    bookkeeper: Option<Bookkeeper>,
+}
+
+impl Named for RegisterUnsharing {
+    fn name() -> &'static str {
+        "register-unsharing"
+    }
+
+    fn description() -> &'static str {
+        "Split apart shared values into separate regsters"
+    }
+}
+
+type RewriteMap<T> = HashMap<T, Vec<(RRC<Cell>, RRC<Cell>)>>;
+
+// A struct for managing the overlapping analysis and rewrite information
+struct Bookkeeper {
+    analysis: ReachingDefinitionAnalysis,
+    widths: HashMap<ir::Id, u64>,
+    group_map: HashMap<ir::Id, RRC<Group>>,
+    cell_map: HashMap<ir::Id, RRC<Cell>>,
+    invoke_map: RewriteMap<ir::Id>,
+}
+
+impl Bookkeeper {
+    // Create a new bookkeeper from the component
+    fn new(comp: &ir::Component) -> Self {
+        // width map is needed to create new registers with the proper widths
+        let widths = comp
+            .cells
+            .iter()
+            .filter_map(|c| {
+                if let ir::CellType::Primitive { name, .. } =
+                    &c.borrow().prototype
+                {
+                    if name == "std_reg" {
+                        if let Some(in_port) = c.borrow().find("in") {
+                            Some((
+                                c.borrow().name.clone(),
+                                in_port.borrow().width,
+                            ))
+                        } else {
+                            None
+                        }
+                    } else {
+                        None
+                    }
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        // (griffin) I'm sorry for the above.
+        // There's probably a cleaner way to write this
+        // TODO(griffin): fix?
+
+        let analysis = ReachingDefinitionAnalysis::new(
+            &comp,
+            &mut comp.control.borrow_mut(),
+        );
+
+        // Used to amortize access to cells and groups that will be needed for
+        // rewriting
+        let group_map = HashMap::new();
```

Comment:
Can you talk a bit more about what this is amortizing exactly? Is it because the `find_cell` and `find_group` methods are linear? The right way to fix that is just to make them use a `LinearHashMap` and have constant access in those methods directly.

Reply:
That's correct. I was trying to avoid messing with stuff beyond the scope of this pass since I'm still learning the existing infrastructure, but I can remove these two in favor of using find_cell and find_group directly.

Can you elaborate on the fix you envision? I can't find documentation on `LinearHashMap`, could you point me in the right direction? Is it in a separate crate from the standard library?

Reply:
I removed the maps and switched to using `find_cell` and `find_group` from the component directly

Reply:
Ah sorry, I meant `LinkedHashMap`. The only important property it has is that it keeps elements in the order of insertion which makes iterating over it deterministic.

Reply:
I see, so you're suggesting changing the the component struct's `groups` and `cells` fields to be `LinkedHashMap`s indexed by Id, rather than `Vec`s?

Reply:
Yeah

Reply:
Do you think it's worth opening a separate issue for and merging this PR, or would it be better to incorporate here? We'll probably have to change a bunch of lines where code is trying to iterate over one/both of these fields

Reply:
Separate PR and issue sounds good